### PR TITLE
sslh 1.18

### PR DESCRIPTION
**Pull request by ilovezfs**
Also,
- remove patch since issue was fixed upstream
- match version in the test

### CODE REVIEW

**Comment by DomT4** [comment]
> set BINDIR=#{bin} to avoid installing in sbin

I'm not massively in love with this until we actually decide some sort of policy on `sbin`, unless other package managers are doing the same thing here.

At the very least, if you're going to start attempting to transition things, to avoid breaking scripts ideally we should be leaving symlinks present in `sbin` for a while.

**Comment by ilovezfs** [comment]
Symlink sounds fine to soften the blow as long as the real one is in `bin` :smirk:

**Comment by ilovezfs** [comment]
However, it's obnoxious that then still triggers the warning.

**Comment by ilovezfs** [comment]
I think I prefer it without the symlink unless we're going to shut the warning up when the target of the symlink is indeed in the PATH.

**Comment by MikeMcQuaid** [comment]
> I'm not massively in love with this until we actually decide some sort of policy on sbin, unless other package managers are doing the same thing here.

I think it would be good to have a policy here but I don't think it should be dependent on other package managers almost all of which requiring running as root/support setuid binaries so `superuser binaries` makes sense for them.

**Comment by ilovezfs** [comment]
@MikeMcQuaid @DomT4 This shuts it up: https://github.com/Homebrew/brew/pull/323

**Comment by DomT4** [comment]
> but I don't think it should be dependent on other package managers almost all of which requiring running as root/support setuid binaries so superuser binaries makes sense for them.

Sorry, I evidently did a bad job explaining my initial comment. I'll expand it a bit to clarify.

> I'm not massively in love with this until we actually decide some sort of policy on sbin, unless other package managers are doing the same thing here.

I'm not massively in love with this until we actually decide some sort of policy on `sbin`, but I'll accept jumping that shark for this particular formula if it has become the norm in other package managers to install _this_ into `bin` rather than `sbin` despite the upstream "default".

**Comment by ilovezfs** [comment]
This doesn't need to be so complicated. /usr/local/bin is a standard path that is in everyone's PATH by default. /usr/local/sbin is not. So unless there's a really good reason specific to a given formula that has been articulated by upstream or technically justifiable with evidence, we should use the standard path.

**Comment by ilovezfs** [comment]
I've removed the symlink because I think that's a bad idea.

**Comment by DomT4** [comment]
> we should use the standard path.

Except as far as upstreams are concerned, they expect their executable to live in `sbin` or they wouldn't be using it. Maybe you'd like to volunteer to "educate" upstream developers, but I think it's a bad idea to arbitrarily deviate from upstream documentation highlighting they expect their executable to live in `sbin`.

And I note, Homebrew already has form on being inconsistent with this particular formula. Adam V moved it from `sbin` to `bin` in 2013, and then Alex moved it back to `sbin` in 2015. That's the kind of thing that hacks off people and shows why we should adopt policy rather than piecemeal.

File: Formula/sslh.rb [hunk 1]
```diff
@@ -20,16 +13,16 @@ class Sslh < Formula
...
-
   depends_on "libconfig"
 
   def install
-    ENV.j1
-    system "make", "install", "PREFIX=#{prefix}"
+    ENV.deparallelize
+    system "make", "install", "PREFIX=#{prefix}", "BINDIR=#{bin}"
+
+    prefix.install "README.MacOSX"
```
Comment (DomT4) [inline Formula/sslh.rb]: FWIW, this file literally only tells people to install `sslh` with MacPorts and offers up an example `launchctl` plist. We can write the latter into the formula ourselves.

**Comment by ilovezfs** [comment]
I consider this an integration issue for OS X when the chosen prefix is /usr/local. If we were on Linux and the prefix was / or /usr, I'd have no problem /sbin or /usr/sbin. We're not on Linux. The prefix is not /. The prefix is not /usr. Nor can it be now that SIP is in place. Therefore, unless Apple adds /usr/local/sbin to the default set of paths, I consider it a serious packaging bug to be putting stuff in non-standard locations, especially when the sole justification is an attribution of intentionality to upstream defaults, even when upstream deliberately makes them configurable.

File: Formula/sslh.rb (hunk 1 above)
Comment (ilovezfs) [inline Formula/sslh.rb]: Yes, I read it.

**Comment by DomT4** [comment]
> I consider this an integration issue for OS X when the chosen prefix is /usr/local

So you'll be adding logic to handle the situation for all our users _not using_ `/usr/local` then, I presume? 😉. Bearing in mind Homebrew may very well not live in `/usr/local` forever it's going to look pretty silly if we move & a year down the line our justification suddenly dies out.

I don't think it makes sense to implement a system on a _per-formula basis_ where you can expect these executables to live in `sbin` on Linux, MacPorts, probably Fink, FreeBSD, and then you start using Homebrew and suddenly that executable lives in a different subdirectory.

**Comment by ilovezfs** [comment]
No special logic is required for non-standard prefixes. They've already added foo_prefix/"bin" to their PATH. If they've also added foo_prefix/"sbin", then it's a matter of indifference for them.

If we move prefixes elsewhere, then it's a question of whether we're adding two paths or one. Not as interesting of an issue in that case, but one is still simpler than two.

As for what happens on Linux, FreeBSD, or MacPorts I think is irrelevant, and I doubt anyone would even notice the difference, but if they did, it be because Linux, et. al were using /sbin or /usr/sbin, not /usr/local/sbin, so it differs regardless.

**Comment by MikeMcQuaid** [comment]
If we're to have any `bin` and `sbin` distinction the one that makes sense in Homebrew's context is having `sbin` those that always require root to run them e.g. `mtr`.

**Comment by ilovezfs** [comment]
Let's just leave it the way it is, since it means less work for us, other than supporting users who get confused when something isn't in their default PATH.

**Comment by MikeMcQuaid** [comment]
@ilovezfs We'll leave it as-is for this PR and I'll make a proposal at some point to detail the version in my comment if you're 👍 on it

**Comment by DomT4** [comment]
> If we're to have any bin and sbin distinction the one that makes sense in Homebrew's context is having sbin those that always require root to run them e.g. mtr.

This would be fine with me, personally.

**Comment by ilovezfs** [comment]
I think we should just leave it as upstream defaults unless there's some specific reason for a particular formula. There are plenty of system binaries that do not require root on OS X. (e.g., diskutil)

**Comment by ilovezfs** [comment]
```
bash-3.2$ which diskutil
/usr/sbin/diskutil
bash-3.2$
```

**Comment by ilovezfs** [comment]
As for things that always require root, yes upstreams will usually be targeting sbin for those if they're not already, and we think they should, we can always ask them in an individual case why they're not.

Of course, if I were they, and someone showed up questioning why I wasn't using sbin for this or that given that it always requires root, I would refer her to
```
$ which sudo
/usr/bin/sudo
```

**Comment by MikeMcQuaid** [comment]
> /sbin is a standard subdirectory of the root directory in Linux and other Unix-like operating systems that contains executable (i.e., ready to run) programs. They are mostly administrative tools, that should be made available only to the root (i.e., administrative) user.

http://www.linfo.org/sbin.html

> `/sbin` ... for scripts with superuser (root) privileges required.

http://askubuntu.com/questions/308045/differences-between-bin-sbin-usr-bin-usr-sbin-usr-local-bin-usr-local/308048#308048

This seems to be most consistent with the system above.

**Comment by MikeMcQuaid** [comment]
Also worth nothing `/etc/paths`:
```
/usr/local/bin
/usr/bin
/bin
/usr/sbin
/sbin
```

**Comment by ilovezfs** [comment]
It's probably worth filing an upstream bug report about if we feel strongly that a particular utility should be targeting sbin not bin. I can't imagine there are many where that actually applies that won't already have an sbin default.

**Comment by xu-cheng** [comment]
It worth noting that:
> /usr/sbin Non-essential system binaries, e.g., daemons for various network-services.
> https://en.wikipedia.org/wiki/Filesystem_Hierarchy_Standard

Also, I share the same opinion among other maintainers that this is a thing should be decided by upstream rather than us. Why to complicate formula, when previous a `./configure --prefix=<prefix>; make install` work just as upstream intended.

**Comment by MikeMcQuaid** [comment]
@xu-cheng Do you think upstreams are considering the default PATH for the default Homebrew installation location on OS X? I don't think they are nor should they.

**Comment by ilovezfs** [comment]
On OS X there is no consistency about what "system" utilities require root. Generally there's a lot more stuff that requires root in /sbin than in /usr/sbin, but there are plenty of utilities that sometimes require root in /bin and /usr/bin, though almost nothing that always requires root.

@xu-cheng only reason would be convenience due to the /etc/paths issue, which is definitely not on upstreams' minds.

If we ever move prefixes, the relevance of the default contents of /etc/paths goes out the window, but we could always add a file to /etc/paths.d/foo if we wanted our new default prefix/"bin" and prefix/"sbin" to be included. That is of course another option here. We could improve UX by having the brew installer do

```
mkdir -p /etc/paths.d
echo /usr/local/sbin > /etc/paths.d/homebrew
```

**Comment by xu-cheng** [comment]
> Do you think upstreams are considering the default PATH for the default Homebrew installation location on OS X? I don't think they are nor should they.

In comparison, on a fresh Linux machine, you still need set this PATH.

```
[root@bot ~]# echo $PATH
/usr/local/bin:/usr/bin
```

Even more, since lots of binaries on sbin are for daemons. They are  never mean to be called by users directly but by `init`.

**Comment by ilovezfs** [comment]
@xu-cheng That's 100% distro dependent:
```
if [ "$EUID" = "0" ] || [ "$USER" = "root" ] ; then
        PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:${ROOTPATH}"
else
        PATH="/usr/local/bin:/usr/bin:/bin:${PATH}"
fi
```

**Comment by MikeMcQuaid** [comment]
@xu-cheng Generally on Linux distros you need to set it as a non-root user but not as a root user.

**Comment by ilovezfs** [comment]
Note that in the example above it's EUID not UID:
```
openzfs@openzfs ~ $ bash -c 'echo $PATH'
/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/x86_64-pc-linux-gnu/gcc-bin/4.9.3:/usr/games/bin
openzfs@openzfs ~ $ sudo bash -c 'echo $PATH'
/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin:/opt/bin:/opt/Citrix/ICAClient:/opt/freerapid:/opt/blackdown-jdk-1.4.2.03/bin:/opt/blackdown-jdk-1.4.2.03/jre/bin:/opt/dmd/bin:/usr/lib/plan9/bin:/opt/nvidia-cg-toolkit/bin:/opt/tivoli/tsm/client/ba/bin:/opt/eagle/bin:/opt/vmware/bin:/opt/aws-as-tools/bin:/opt/aws-cw-tools/bin:/opt/aws-elb-tools/bin:/opt/aws-iam-tools/bin:/opt/aws-rds-tools/bin:/usr/bin/cdsclient:/opt/cuda/bin:/opt/ekopath/bin:/usr/ti-linux-gnu/tigcc-bin/4.1.2:/usr/ti-linux-gnu/bin
```

**Comment by ilovezfs** [comment]
```
openzfs@openzfs ~ $ which zfs
which: no zfs in (/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/x86_64-pc-linux-gnu/gcc-bin/4.9.3:/usr/games/bin)
openzfs@openzfs ~ $ sudo which zfs
/sbin/zfs
openzfs@openzfs ~ $
```

**Comment by ilovezfs** [comment]
@DomT4 :shipit: ?

**Comment by DomT4** [comment]
Aye 👍. We've got at least a few `sbin` conversations ongoing, it'd be nice to centralise them somewhere but a formula PR isn't the ideal place.

**Comment by ilovezfs** [comment]
@DomT4 I think we should just treat the sbin cases where we "disagree" with upstream's choice for whatever reason (a game in sbin, a system utility with no non-root use cases in bin, etc.) as upstream bug report territory. If we change it prospectively, comment with issue link. Or wait for them to do it. Just like any other bug. :)

**Pull request closed without merging**